EZP-30988: Implemented Section filtering URL Query Criterion

ITernovtsiy opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30988](https://jira.ez.no/browse/EZP-30988)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `7.5.x` for eZ Platform `v2.5.x` LTS
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

- Implement Section criteria for URL lookup
- Fix slow query when filter by VisibleOnly criteria

// Edited by @alongosz:
### QA
- You can play with the test command to find URLs filtered by Content Section:
https://github.com/ezsystems/ezplatform/compare/2.5...alongosz:for-qa-ezp-30988-find-urls-cmd.
Usage: `php ./bin/console test:find-urls [<url1> <url2> ...]`. No arguments returns all registered URLs. An URL is registered when e.g. added to Content RichText field or set via URL Field. By manipulating Content item Section you can test the feature.
- Sanity checks for LinkManager (might be an overkill, because new code is not used there).

### Doc

I can't find doc about API to find URLs. Either it's too late and I'm not seeing something obvious or indeed we need to document the whole thing (more than the scope of this feature).

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

ITernovtsiy commented:
tests failed in untouched code, all related tests pass

adamwojs commented:
Hi @ITernovtsiy! Thank you for contribution, I will review changes later. 

tests failed in untouched code, all related tests pass

> tests failed in untouched code, all related tests pass

Solr tests should be fixed by https://github.com/ezsystems/ezpublish-kernel/pull/2784 (rebase is required)

ITernovtsiy commented:
rebased

mikadamczyk commented on eZ/Publish/API/Repository/Tests/Values/URL/Query/Criterion/SectionTest.php:
```diff
@@ -0,0 +1,26 @@
+     *
+     * @covers \eZ\Publish\API\Repository\Values\Content\Section::__construct
+     */
+    public function testNewClass()
```
```suggestion
    public function testInstanceCreation()
```

ITernovtsiy replied:
@mikadamczyk
There are 10 other functions named as `testNewClass`, that was the reason I picked it.
May I rename them all or just this one? Or may I leave it as it is?

mikadamczyk replied:
It's not critical to me, especially if we have this name already. It can stay the way it is @ITernovtsiy

ITernovtsiy commented:
@adamwojs Yes, we're using this with `v2.5` and will be able to clean-up project code a little bit (if this can be merged into 2.5).
If not, v3 is okay. While we have this on one project only, it's not so critical :)
Let me know if I should rebase it with master. 
Thanks.

ITernovtsiy commented on eZ/Publish/Core/Persistence/Legacy/URL/Query/CriterionHandler/Base.php:
```diff
@@ -0,0 +1,43 @@
+{
+    /**
+     * Inner join `ezurl_object_link` table if not joined yet.
+     * @param SelectQuery $query
```
Do we need to use FQN in PHPDoc while we have the class imported?
I see short names are used commonly in `ezpublish-kernel`.

mikadamczyk replied:
Yes, now we use FQCN

SerheyDolgushev replied:
@mikadamczyk sorry, what are the reasons for that? Is it going to change at some point?

alongosz replied:
FQCN in PhpDocs is used for readability purposes. There are many class, especially in our ecosystem, which share the same name (e.g. `Repository`). I need to see during review that type hint is correct (for API layer, not for something else). We will automate FQCN checks in the future.

> Is it going to change at some point?

We gradually move away from documenting strict types, so if anything is strict, you can skip doc for it (but e.g. `array` is not considered strict, so be careful).

Besides, comments are not part of executable code. Imports are (I know, not exactly, but still...). Even though IDE handles that, it's still a piece of documentation and w/o FQCN some information is missing.

ITernovtsiy commented on eZ/Publish/Core/Persistence/Legacy/Tests/URL/Query/CriterionHandler/SectionTest.php:
```diff
@@ -0,0 +1,129 @@
+use eZ\Publish\Core\Persistence\Legacy\URL\Query\CriteriaConverter;
+use eZ\Publish\Core\Persistence\Legacy\URL\Query\CriterionHandler\Section as SectionHandler;
+
+class SectionTest extends CriterionHandlerTest
```
implemented `testFindUrlsUsingSectionCriterion` and `testFindUrlsUsingSectionAndValidityCriterionValid`. Let me know if this `SectionTest` should be removed

lserwatka commented:
We can bring this to 2.5.

ITernovtsiy commented:
Thanks for review @alongosz 
I fixed code style, removed unit tests, renamed `Section` to be `SectionId` and introduced `SectionIdentifier` criteria. And, added more integration tests to cover them.
Please review once again and let me know if I missed anything.

adamwojs commented on eZ/Publish/API/Repository/Values/URL/Query/Criterion/SectionIdentifier.php:
```diff
@@ -0,0 +1,30 @@
+    public $sectionIdentifiers;
+
+    /**
+     * @param int[] $sectionIdentifiers
```
```suggestion
     * @param string[] $sectionIdentifiers
```

ITernovtsiy replied:
fixed

lserwatka commented:
We will ship it in 2.5.7💪👌🎉well done to all involved!

micszo commented:
Testing discovered: https://jira.ez.no/browse/EZP-31059.
Also that issue affects this feature (but waiting for root cause).

![EZP-31059 2](https://user-images.githubusercontent.com/13622502/67197968-3d779280-f3fe-11e9-82e8-d2591437d609.png)

lserwatka commented:
@alongosz could you merge it up?

alongosz commented:
@lserwatka Done.

Thank you @ITernovtsiy :tada:
